lib/bolt/executor.rb

```diff
@@ -373,83 +382,156 @@ def run_plan(scope, plan, params)
-            raise Bolt::InvalidParallelResult.new(result.to_s, *Puppet::Pops::PuppetStack.top_of_stack)
-          end
+        result
+      rescue Puppet::PreformattedError => e
+        if e.cause.is_a?(Bolt::Error)
+          e.cause
+        else
+          raise e
+        end
+      end
 
-          result
-        rescue Puppet::PreformattedError => e
-          if e.cause.is_a?(Bolt::Error)
-            e.cause
+      fiber = Bolt::Future.new(fiber, name)
+      @logger.trace("Created future #{name}") if in_parallel
+      plan_fibers.fibers << fiber
+      fiber
+    end
+
+    # Check on all the Fibers and see if any can be resumed.
+    def round_robin(rrfibers = plan_fibers.all)
+      r = nil
+
+      rrfibers.each do |fiber|
+        # If the Fiber is still running and can be resumed, then resume it and
+        # save the output to check if the task has finished or not.
+        @logger.trace("Checking future '#{fiber.name}'") if in_parallel
+        if fiber.alive?
+          @logger.trace("Resuming future '#{fiber.name}'") if in_parallel
+          # Stop any spins that might have started from a previous action so we
+          # don't get multiple spinners.
+          publish_event(type: :stop_spin)
+          r = fiber.resume
+        # If the Fiber is dead that means it's returned a result, which we
+        # automatically store on the Bolt::Future object when we resume it
+        # (since it can't be retrieved once the Fiber dies).
+        else
+          # Reset the sleep since something successfully resumed
+          @sleep = 0.1
+          @logger.trace("Retrieving result from future '#{fiber.name}'") if in_parallel
+          # Get the result from the Bolt::Future object
+          # Remove the fiber
+          if fiber == main_plan
+            fiber_results.main_plan = fiber.value
+            plan_fibers.main_plan = nil
           else
-            raise e
+            fiber_results.fibers[fiber.name] = fiber.value
+            plan_fibers.fibers.delete(fiber)
           end
         end
       end
 
-      Bolt::Yarn.new(fiber, index)
-    end
+      return unless r.is_a?(Symbol) && r == :unfinished
+      @logger.trace("Nothing can be resumed. Rechecking in #{@sleep} seconds.") if in_parallel
 
-    def handle_event(event)
-      case event[:type]
-      when :node_result
-        @thread_completed = true
-      end
+      # Rather than trying
+      @sleep = @sleep > 1 ? @sleep : @sleep * 2
+      sleep(@sleep)
     end
 
-    def round_robin(skein)
-      subscribe(self, [:node_result])
-      results = Array.new(skein.length)
-      @in_parallel = true
-      publish_event(type: :stop_spin)
+    def plan_complete?
+      plan_fibers.fibers.empty? && main_plan.nil?
+    end
 
-      until skein.empty?
-        @thread_completed = false
-        r = nil
+    def wait(fibers, timeout, catch_errors: false, **_kwargs)
```

Comment:
I think we should already be round_robining so we don't need to call round_robin in this. We could just loop over all the provided fibers checking if they're alive but this will be much easier if each  `Bolt::Future` contained a concurrent ruby future we could check once without polling.